Desktop/widgets/sourceitem.cpp

```diff
@@ -0,0 +1,406 @@
+		, const JASPListControl::LabelValueMap&		values
+		, QAbstractItemModel*						nativeModel
+		, const QVector<SourceItem*>&				discardSources
+		, const QVector<QMap<QString, QVariant> >&	conditionVariables
+		) : _listControl(listControl)
+{
+	_name					= map["name"].toString();
+	_controlName			= map["controlName"].toString();
+	_modelUse				= map["use"].toString();
+	_conditionExpression	= map["condition"].toString();
+	_values					= values;
+	_nativeModel			= nativeModel;
+	_discardSources			= discardSources;
+
+	_isValuesSource			= map.contains("isValuesSource")			? map["isValuesSource"].toBool()			: false;
+	_isDataSetColumns		= map.contains("isDataSetColumns")			? map["isDataSetColumns"].toBool()			: false;
+	_combineWithOtherModels	= map.contains("combineWithOtherModels")	? map["combineWithOtherModels"].toBool()	: false;
+	_nativeModelRole		= map.contains("nativeModelRole")			? map["nativeModelRole"].toInt()			: Qt::DisplayRole;
+
+	if (!_controlName.isEmpty()) _usedControls.insert(_controlName);
+
+
+	for (const QMap<QString, QVariant>& conditionVariable : conditionVariables)
+	{
+		_conditionVariables.push_back(ConditionVariable(conditionVariable["name"].toString()
+									, conditionVariable["component"].toString()
+									, conditionVariable["property"].toString()
+									, conditionVariable["addQuotes"].toBool())
+					);
+		if (!conditionVariable["component"].toString().isEmpty()) _usedControls.insert(conditionVariable["component"].toString());
+	}
+
+	_setUp();
+}
+
+SourceItem::SourceItem(JASPListControl *listControl, const JASPListControl::LabelValueMap &values)
+	:  _listControl(listControl), _values(values), _isValuesSource(true)
+{
+	_setUp();
+}
+
+SourceItem::SourceItem(JASPListControl *listControl, QAbstractItemModel *nativeModel, int nativeModelRole)
+	 : _listControl(listControl), _nativeModel(nativeModel), _nativeModelRole(nativeModelRole)
+{
+	_setUp();
+}
+
+SourceItem::SourceItem(JASPListControl *listControl)
+	:  _listControl(listControl), _isDataSetColumns(true)
+{
+	_setUp();
+}
+
+
+void SourceItem::_setUp()
+{
+	ListModel* sourceModel = nullptr;
+
+	if (_isDataSetColumns)
+	{
+		_nativeModel = AnalysisForm::columnsModel;
+		_nativeModelRole = AnalysisForm::columnsModelRole;
+	}
+
+	if (_isValuesSource)								sourceModel = new ListModelLabelValueTerms(_listControl, _values);
+	else if (_nativeModel)								sourceModel = qobject_cast<ListModel*>(_nativeModel); // A nativeModel is not always a ListModel
+	else if (_listControl->form() && !_name.isEmpty())	sourceModel = _listControl->form()->getModel(_name);
+
+	if (sourceModel)
+	{
+		_model = sourceModel;
+		_listControl->addDependency(_model->listView());
+		ListModel::connect(_model, &ListModel::termsChanged, _listControl->model(), &ListModel::sourceTermsChanged);
+	}
+	else if (_nativeModel)
+	{
+		connect(_nativeModel, SIGNAL(dataChanged(const QModelIndex &, const QModelIndex &, const QVector<int> &)),	this,	SLOT(nativeModelChanged()) );
```

Comment:
New style `connect`'s weren't possible I suppose?

Reply:
(You can refer to functions like `&SuperClass::Function`)

Reply:
Because of the parameters, I chose this case. But I could use a lambda instead, and adding the sourceItem as context to be sure that the connection is destroyed when the SourceItem is destroyed.

Reply:
But the normal connect also drops the parameters right?
As in, I don't see why this version would be better?

Reply:
By that I mean, why not do:
`connect(_nativeModel, &QAbstractItemModel::dataChanged, this, &SourceItem::nativeModelChanged);` should work I think.

Reply:
That's true. At some point I thought that would not work, but it works indeed!

Reply:
Yeah, connects automatically drop arguments/parameters from the signal if they are not needed/accepted by the slot.